### PR TITLE
`@0x/contracts-integrations`: Better input generation for fuzzing

### PR DESCRIPTION
**Pull request by mzhu25**
## Description
Adds some functionality to `Pseudorandom` for more sophisticated input generation. In particular, you can now pass in an array of weights alongside the array of inputs to `Pseudorandom.sample` and `sampleSize`. For `Pseudorandom.integer`, you can pass in a custom probability distribution alongside the min and max values. I implemented the [Kumaraswamy distribution](https://www.johndcook.com/blog/2009/11/24/kumaraswamy-distribution/), a beta-like distribution which allows to weigh the extreme ends of an input distribution more heavily: 
Screen Shot 2019-12-13 at 13 17 52
I want to get some feedback on what sort of weights/distributions we want. I've chosen some values that are intended to be either more realistic or hit more corner cases than uniformly random samples, but they're still p arbitrary. 

## Checklist:

-   [ ] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.
-   [ ] Update documentation as needed.
-   [ ] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

File: contracts/integrations/test/fuzz_tests/stake_management_test.ts [hunk 1]
```diff
@@ -20,14 +21,15 @@ export class StakeManagementSimulation extends Simulation {
...
-            ...stakers.map(staker => staker.simulationActions.validUnstake),
-            ...stakers.map(staker => staker.simulationActions.validMoveStake),
-            poolManagement.generator,
-        ];
+        const [actions, weights] = _.unzip([
+            ...stakers.map(staker => [staker.simulationActions.validStake, 0.3 / stakers.length]),
+            ...stakers.map(staker => [staker.simulationActions.validUnstake, 0.2 / stakers.length]),
+            ...stakers.map(staker => [staker.simulationActions.validMoveStake, 0.3 / stakers.length]),
+            [poolManagement.generator, 0.2],
+        ]) as [Array<AsyncIterableIterator<AssertionResult | void>>, number[]];
```
Comment (mzhu25) [inline contracts/integrations/test/fuzz_tests/stake_management_test.ts]: anyone have a cleaner way to do this? 🤔

Comment (hysz) [reply contracts/integrations/test/fuzz_tests/stake_management_test.ts]: @dorothy-zbornak 😬

File: contracts/integrations/test/fuzz_tests/pool_membership_test.ts [hunk 2]
```diff
@@ -22,14 +23,14 @@ export class PoolMembershipSimulation extends Simulation {
 
         const poolManagement = new PoolManagementSimulation(this.environment);
 
-        const actions = [
-            ...makers.map(maker => maker.simulationActions.validJoinStakingPool),
-            ...takers.map(taker => taker.simulationActions.validFillOrder),
-            poolManagement.generator,
-        ];
+        const [actions, weights] = _.unzip([
+            ...makers.map(maker => [maker.simulationActions.validJoinStakingPool, 0.2 / makers.length]),
```
Comment (hysz) [inline contracts/integrations/test/fuzz_tests/pool_membership_test.ts]: Could be good to make these constants to add context around `0.2` and `0.6` below.

**Comment by hysz** [comment]
@mzhu25 Will this distribution allow us to generate high density around the boundaries. Or perhaps there is a someway that we can increase probability density around a custom set of points defined by the user? Could also be interesting to swap out this distribution for another.

File: contracts/integrations/test/framework/actors/pool_operator.ts [hunk 3]
```diff
@@ -83,7 +83,11 @@ export function PoolOperatorMixin<TBase extends Constructor>(Base: TBase): TBase
         private async *_validCreateStakingPool(): AsyncIterableIterator<AssertionResult> {
             const assertion = validCreateStakingPoolAssertion(this.actor.deployment, this.actor.simulationEnvironment!);
             while (true) {
-                const operatorShare = Pseudorandom.integer(constants.PPM).toNumber();
+                const operatorShare = Pseudorandom.integer(
+                    0,
+                    constants.PPM,
+                    Distributions.Kumaraswamy(0.2, 0.2),
```
Comment (hysz) [inline contracts/integrations/test/framework/actors/pool_operator.ts]: Can we document the `(0.2, 0.2)`?

File: contracts/integrations/test/framework/actors/staker.ts [hunk 4]
```diff
@@ -118,7 +118,10 @@ export function StakerMixin<TBase extends Constructor>(Base: TBase): TBase & Con
                 const fromStatus =
                     fromPoolId === undefined || stakingPools[fromPoolId].lastFinalized.isLessThan(currentEpoch.minus(1))
                         ? StakeStatus.Undelegated
-                        : (Pseudorandom.sample([StakeStatus.Undelegated, StakeStatus.Delegated]) as StakeStatus);
+                        : (Pseudorandom.sample(
+                              [StakeStatus.Undelegated, StakeStatus.Delegated],
+                              [0.2, 0.8],
```
Comment (hysz) [inline contracts/integrations/test/framework/actors/staker.ts]: Here as well, along w others 🙏

File: contracts/integrations/test/framework/utils/pseudorandom.ts [hunk 5]
```diff
@@ -1,54 +1,80 @@
...
+            const x = this.rng();
+            index = cdf.findIndex(value => value > x);
+        } else {
+            index = Math.abs(this.rng.int32()) % arr.length;
+        }
         return arr[index];
     }
 
     /*
      * Pseudorandom version of _.sampleSize. Returns an array of `n` samples from the given array
```
Comment (hysz) [inline contracts/integrations/test/framework/utils/pseudorandom.ts]: Let's add a note about the weights here.

File: contracts/integrations/test/fuzz_tests/stake_management_test.ts (hunk 1 above)
Comment (dorothy-zbornak) [reply contracts/integrations/test/fuzz_tests/stake_management_test.ts]: I honestly don't think it's that bad. Maybe just comment what's going on?

File: contracts/integrations/test/framework/actors/pool_operator.ts (hunk 3 above)
Comment (dorothy-zbornak) [reply contracts/integrations/test/framework/actors/pool_operator.ts]: How about having `0.2` be the default `alpha` and `beta` values and documenting them in the devdoc.

**Comment by mzhu25** [comment]
@hysz @dorothy-zbornak here's the pdf for (0.2, 0.2):
Screen Shot 2019-12-17 at 20 14 13

File: contracts/integrations/test/fuzz_tests/staking_rewards_test.ts [hunk 6]
```diff
@@ -32,15 +33,20 @@ export class StakingRewardsSimulation extends Simulation {
...
 
-        const actions = [
-            ...stakers.map(staker => staker.simulationActions.validWithdrawDelegatorRewards),
-            ...keepers.map(keeper => keeper.simulationActions.validFinalizePool),
-            ...keepers.map(keeper => keeper.simulationActions.validEndEpoch),
-            poolMembership.generator,
-            stakeManagement.generator,
-        ];
+        const [actions, weights] = _.unzip([
+            // 10% chacne of executing validWithdrawDelegatorRewards for a random staker
```
Comment (hysz) [inline contracts/integrations/test/fuzz_tests/staking_rewards_test.ts]: 🇨🇦

File: contracts/integrations/test/framework/actors/staker.ts [hunk 7]
```diff
@@ -128,15 +131,18 @@ export function StakerMixin<TBase extends Constructor>(Base: TBase): TBase & Con
                 const toStatus =
                     toPoolId === undefined || stakingPools[toPoolId].lastFinalized.isLessThan(currentEpoch.minus(1))
                         ? StakeStatus.Undelegated
-                        : (Pseudorandom.sample([StakeStatus.Undelegated, StakeStatus.Delegated]) as StakeStatus);
+                        : (Pseudorandom.sample(
+                              [StakeStatus.Undelegated, StakeStatus.Delegated],
+                              [0.2, 0.8], // 20$ chance of `Undelegated`, 80% chance of `Delegated`
```
Comment (hysz) [inline contracts/integrations/test/framework/actors/staker.ts]: 🇨🇦
